src/communication/websocket/session.cpp

```diff
@@ -88,13 +91,43 @@ void Session::DoRead() {
       }));
 }
 
-void Session::OnRead(boost::beast::error_code ec, size_t /*bytes_transferred*/) {
+bool Session::Authenticate() {
+  try {
+    const auto creds = nlohmann::json::parse(boost::beast::buffers_to_string(buffer_.data()));
+    if (auth_.Authenticate(creds.at("username").get<std::string>(), creds.at("password").get<std::string>())) {
+      authenticated_ = true;
+      return true;
+    }
+  } catch (const nlohmann::json::out_of_range &out_of_range) {
+    spdlog::error("Invalid JSON for authentication received: {}!", out_of_range.what());
+  } catch (const nlohmann::json::parse_error &parse_error) {
+    spdlog::error("Cannot parse JSON for WebSocket authentication: {}!", parse_error.what());
+  }
+  return false;
+}
+
+void Session::OnRead(const boost::beast::error_code ec, const size_t /*bytes_transferred*/) {
   if (ec == boost::beast::websocket::error::closed) {
     messages_.clear();
     connected_.store(false, std::memory_order_relaxed);
     return;
   }
 
+  if (!authenticated_) {
```

Comment:
You need to check before Write if it's authenticated.

Reply:
What do you mean before `Write`? I am checking if the user is authenticated and if not I try to authenticate him, is there something wrong here?

Reply:
You don't want to send the logs to the user that isn't authorized and authenticated for that.
So Write should check if the user is authorized before writing anything on the socket.

Reply:
Aha ok then I need to add the same logic to the Write method, I was confused why you commented this here?

Reply:
Sorry didn't see a better place to write it as you didn't modify the method.